docs/versioning.md

```diff
@@ -0,0 +1,101 @@
+
+While the versioning mechanisms are clearly specified in this document, the policies for version management are only mandated for IR version and operator verison. 
+
+For model version, ONNX users and systems MAY follow whatever local customs make sense, however, to facilitate shared libraries or repositories of ONNX models to be managable, models SHOULD adhere to the policies described under Model versioning 
+
+
+
+### SemVer, Files and Frameworks
+
+ONNX builds on the principles and syntax defined by [SemVer 2.0.0](http://semver.org/spec/v2.0.0.html). Throughout this document, we use the terms *breaking change*, *non-breaking change*, and *patch* consistent with SemVer 2.0.0. 
+
+The SemVer specification is written in terms of the dependencies between the caller and callee of a given API.  With respect to operator definitions, the NodeProto represents the caller, and the operator specification represents the callee.
+
+Because ONNX models are serialized files, not APIs, it's worth making clear how the dependency between a serialized model and a piece of software that consumes that model relate.  As a rough approximation, the serialized model plays the role of an API's *callee*; the consumer of the seriaized model plays the role of the API's *caller.*
+
+The ONNX versioning principles are based on [Postel's law](https://en.wikipedia.org/wiki/Robustness_principle): be conservative in what you do, be liberal in what you accept from others.  
+
+1. A producer of a given ONNX model (and the ONNX specification itself) MUST strictly adhere to the rules for breaking vs. non-breaking changes defined in this specification.
+2. A consumer of a given ONNX model SHOULD consume an updated ONNX file provided there are no breaking changes in the new ONNX file's IR version, referenced operator versions, or model version (e.g., the MAJOR version numbers are have not changed between the two ONNX files) .
+3. A consumer of a given ONNX model MAY consume an updated ONNX file provided there are one or more breaking changes in the new ONNX file's IR version, referenced operator versions, or model version.
+
+Per the rules of SemVer 2.0, during the initial development of ONNX, we use a MAJOR version of 0 for both the IR version and operator version and will only increment the MINOR version in the face of either a breaking change or the need to stabiize a version for specific engineering purposes. [ISSUE: decide how we will stabilize and archive released versions of ONNX] 
+
+### Serializing SemVer version numbers in protobuf
+
+For historical reasons, ONNX serializes the MAJOR, MINOR, and PATCH values as a bit-packed 32-bit integer; the most siginificant byte is the MAJOR component, the second most significant byte is the MINOR component, the least significant two bytes are the PATCH component. 
```

Comment:
Haha, "for historical reasons." One good reason for representing this using a bit-packed 32-bit integer is that it's modestly simpler to implement the comparison.

This raises a question in my mind, however. Suppose we release ONNX IR version 0.0.10 (in decimal). Are we representing this as `0x00000010` or `0x0000000A`?  Better be the latter, right? In that case, will we conventionally refer to the IR version in hex or decimal? Maybe the IP address conventions will be unambiguous enough...

Also, one last bit of snark: "65536 ought to be more than enough for anyone." ;)

Reply:
Yes, it's the latter. I updated the example to be 1.2.345 -> 0x01020157

And yes, it's like IP addresses only there are three parts, not 4 (as in IPv4).

Reply:
I love using a 32-bit int btw. It's just we are going to need a place to stash the release and build metadata info.  For example, I just noticed that we've done two releases (v0.1 and v0.2) yet there's no way to identify whether a PB came from v0.1 or v0.2.